torchvision/models/detection/_utils.py

```diff
@@ -274,17 +279,25 @@ def __init__(self, high_threshold, low_threshold, allow_low_quality_matches=Fals
         self.low_threshold = low_threshold
         self.allow_low_quality_matches = allow_low_quality_matches
 
-    def __call__(self, match_quality_matrix):
+        # if box_similarity is None:
+        #    box_similarity = box_ops.box_iou
+        # self.box_similarity = box_similarity
+
+    def __call__(self, gt_boxes, anchors_per_image):
         """
         Args:
-            match_quality_matrix (Tensor[float]): an MxN tensor, containing the
-            pairwise quality between M ground-truth elements and N predicted elements.
+            gt_boxes (Tensor[float]): an Mx4 tensor, containing M detections.
+
+            anchors_per_image (Tensor[float]): an Mx4 tensor, containing
+                the anchors for a specific image.
 
         Returns:
             matches (Tensor[int64]): an N tensor where N[i] is a matched gt in
             [0, M - 1] or a negative value indicating that prediction i could not
             be matched.
         """
+        match_quality_matrix = box_ops.box_iou(gt_boxes, anchors_per_image) #  self.box_similarity(gt_boxes, anchors_per_image)
```

Comment:
Is there a need for this change? The previous implementation seemed more generic, as it didn't assume that the matching was the `box_iou`

Reply:
So exposing it in the constructor like this:
```python
    def __init__(self,
                 high_threshold,
                 low_threshold,
                 allow_low_quality_matches=False,
                 box_similarity=None):
        # type: (float, float, bool)
```

Produces the following runtime error:

```shell
Number of type annotations (3) did not match the number of method parameters (4):
  File "/home/hgaiser/dev/vision/torchvision/models/detection/_utils.py", line 257
    def __init__(self,
    ~~~~~~~~~~~~~~~~~~...  <--- HERE
                 high_threshold,
                 low_threshold,
```

This makes sense because the type hints don't match, but I didn't know that torch actually uses them for runtime error checking. 

According to the [TorchScript documentation](https://pytorch.org/docs/stable/jit.html#supported-type) it is using the [mypy type hints](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#functions), which would mean I should be able to do:

```python
# type: (float, float, bool, Callable[[torch.tensor, torch.tensor], torch.tensor])
```

Which gives me the error :
```shell
Unknown type constructor Callable:
```
(even after importing `Callable` from `typing`)

I'm stuck here because I don't know in what format TorchScript expects a callable.

Currently in `rpn.py` it isn't passed as a constructor argument, but instead created as:
```python
self.box_similarity = box_ops.box_iou
```

But this gives the following error:

```shell
  File "/home/hgaiser/dev/vision/torchvision/models/detection/_utils.py", line 282
        self.low_threshold = low_threshold
        self.allow_low_quality_matches = allow_low_quality_matches
        self.box_similarity = box_ops.box_iou
                              ~~~~~~~~~~~~~~~ <--- HERE
```

I'm assuming what I'm isn't even possible? I feel like I'm missing something conceptually around TorchScript, but I don't know what :). Some pointers would be welcome.

Reply:
Let's take a step back: why can't we pass the similarity matrix to the `Matcher`?

Reply:
Ah I see, I misinterpreted your previous comment. Very good question! I'm sure I had a reason at the time but it was a little while back and I can't recall :')

I'll have a look to see if I can revert the change and if I re-discover the reason I will let you know, but very sharp eye!

Reply:
I've put it back to what it was and I see no issue, must be ghosts then :)